Userland/Libraries/LibWeb/HTML/Worker.cpp

```diff
@@ -0,0 +1,363 @@
+    //      7b. Otherwise, create a new DedicatedWorkerGlobalScope object.
+    // FIXME: Proper support for both SharedWorkerGlobalScope and DedicatedWorkerGlobalScope
+    if (is_shared)
+        TODO();
+
+    // FIXME: Make and use subclasses of WorkerGlobalScope, however this requries JS::GlobalObject to
+    //        play nicely with the IDL interpreter, to make spec-compliant extensions, which it currently does not.
+    m_worker_scope = static_cast<JS::GlobalObject*>(m_worker_vm->heap().allocate_without_global_object<JS::GlobalObject>());
+    m_worker_scope->initialize_global_object();
+
+    m_console = adopt_ref(*new WorkerDebugConsoleClient(m_worker_scope->console()));
+    m_worker_scope->console().set_client(*m_console);
+
+    // FIXME: This should be done with IDL
+    u8 attr = JS::Attribute::Writable | JS::Attribute::Enumerable | JS::Attribute::Configurable;
+    m_worker_scope->define_native_function(
+        "postMessage", [this](auto& vm, auto&) {
+            // This is the implementation of the function that the spawned worked calls
+
+            if (vm.argument_count() < 1)
+                return JS::js_undefined();
+
+            auto message = vm.argument(0);
+            // FIXME: `transfer` not support by post_message yet
+
+            // The postMessage(message, transfer) and postMessage(message, options) methods on DedicatedWorkerGlobalScope
+            // objects act as if, when invoked, it immediately invoked the respective postMessage(message, transfer) and
+            // postMessage(message, options) on the port, with the same arguments, and returned the same return value
+
+            dbgln_if(WEB_WORKER_DEBUG, "WebWorker: Inner post_message");
+
+            // FIXME: This is a bit of a hack, in reality, we should m_outside_port->post_message and the onmessage event
+            //        should bubble up to the Worker itself from there.
+
+            auto& event_loop = get_vm_event_loop(m_document->realm().vm());
+
+            event_loop.task_queue().add(HTML::Task::create(HTML::Task::Source::PostedMessage, nullptr, [this, message]() mutable {
+                MessageEventInit event_init {};
+                event_init.data = message;
+                event_init.origin = "<origin>";
+                dispatch_event(MessageEvent::create(HTML::EventNames::message, event_init));
+            }));
+
+            return JS::js_undefined();
+        },
+        2, attr);
+
+    m_execution_context.current_node = nullptr;
+    m_execution_context.this_value = m_worker_scope;
+    m_execution_context.function_name = "(global execution context)"sv;
+    m_execution_context.lexical_environment = &m_worker_realm->global_environment();
+    m_execution_context.variable_environment = &m_worker_realm->global_environment();
+    m_execution_context.realm = m_worker_realm;
+    m_execution_context.is_strict_mode = true;
```

Comment:
Is strict mode always supposed to be enabled? If so, add a `Note:` why/from where in the spec.

Reply:
This just came from the reference implementation I was using to learn how to setup the jsvm (`Workbook.cpp`)
No other reason than that.

Reply:
For now I just added a `FIXME` stating that it came from a reference implementation, and that it should be replaced with spec-appropriate configuration.

Reply:
> For now I just added a FIXME stating that it came from a reference implementation

What does that mean?! I get extremely sceptical when someone cites other code as their source of truth, that's not how this should work, and we don't want to *copy another implementation*, but *implement a specification*. Please clarify.

Reply:
I don't mean to imply I'm using it as a source-of-truth, simply as a kicking-off point for getting a working execution context setup, that can be brought closer in line to the spec over time. Maybe there is a better way of phrasing the comment to convey that, and the `is_strict_mode` call is probably un-needed.

The spec specifically glosses over multithreading with  an "[up to you](https://html.spec.whatwg.org/multipage/infrastructure.html#in-parallel)" explanation, so makes no mention of setting up an alternate vm or execution context. In reality they likely want you to use the same VM instance but have it support truly parallel operation. This seemed to be the only way to get a script to run independently of the one in the browser at the moment.

Maybe I'm just going about this the wrong way. 😕

Reply:
> simply as a kicking-off point for getting a working execution context setup, that can be brought closer in line to the spec over time.

That's fine, my concern is more taking code from somewhere else than getting everything right in the first revision. What's this "reference implementation" you're talking about? Another engines implementation of workers? :thinking:

> This seemed to be the only way to get a script to run independently of the one in the browser at the moment.

Just creating another VM in the same thread does not magically run things in parallel though :)

Reply:
When I say "reference implementation" I simply am talking about a reference (within serenity) for getting the JS interpreter up and running at a base level. In this case I was basing the setup code off of how `Workbook.cpp` sets up it's JS interpreter.

> Just creating another VM in the same thread does not magically run things in parallel though :)

I was worried that might be the case, but I'm not really familiar enough with all the serenity code to break this out properly into a whole thread. Perhaps it was a bit of a mistake to dive into such an integrated feature from the start 😟

Reply:
> In this case I was basing the setup code off of how Workbook.cpp sets up it's JS interpreter.

Ah, that's totally fine then! New comment is better 👍

> I was worried that might be the case, but I'm not really familiar enough with all the serenity code to break this out properly into a whole thread.

None of the "run in parallel" parts of LibWeb do this correctly yet. Just slap a FIXME on it (as you did) and we'll get to it eventually! See also: https://twitter.com/linusgroh/status/1419426936582021120 :)